### PR TITLE
Add VFS CAS implementation.

### PR DESCRIPTION
**Pull request by ljcollins25**

### CODE REVIEW

File: Public/Src/Cache/ContentStore/Interfaces/FileSystem/VfsUtilities.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            {
+                return false;
+            }
+
+            if (candidateParent.EndsWith(DirectorySeparatorCharString))
+            {
+                return true;
+            }
+
+            return path.Length > candidateParent.Length &&
```
Comment (narasamdya) [inline Public/Src/Cache/ContentStore/Interfaces/FileSystem/VfsUtilities.cs]: >path.Length > candidateParent.Length [](start = 19, length = 36)

`if (path.Length < candidateParent.Length) return false;` can be put at the top, no?

File: Public/Src/Cache/ContentStore/Interfaces/FileSystem/VfsUtilities.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        /// Gets whether a path is contained in another path
+        /// </summary>
+        public static bool IsPathWithin(this string path, string candidateParent)
+        {
+            if (!path.StartsWith(candidateParent, StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            if (candidateParent.EndsWith(DirectorySeparatorCharString))
```
Comment (narasamdya) [inline Public/Src/Cache/ContentStore/Interfaces/FileSystem/VfsUtilities.cs]: Why don't you just append `candidateParent` with `\` if it's not ending with `\`?

File: Public/Src/Cache/ContentStore/Vfs/VfsCasConfiguration.cs [hunk 3]
```diff
@@ -42,6 +42,26 @@ public class VfsCasConfiguration
...
+        /// </summary>
+        public AbsolutePath VfsMountRootPath { get; }
+
+        /// <summary>
+        /// The root of mounted virtualized directories
+        /// </summary>
+        public RelativePath VfsMountRelativeRoot { get; }
+
+        /// <summary>
+        /// The root of mounted virtualized directories
```
Comment (narasamdya) [inline Public/Src/Cache/ContentStore/Vfs/VfsCasConfiguration.cs]: >The root of mounted virtualized directories [](start = 12, length = 43)

Copy-paste?

File: Public/Src/Cache/ContentStore/Vfs/VfsCasConfiguration.cs [hunk 4]
```diff
@@ -42,6 +42,26 @@ public class VfsCasConfiguration
         /// </summary>
         public AbsolutePath VfsRootPath { get; }
 
+        /// <summary>
+        /// The root of mounted virtualized directories
+        /// </summary>
+        public AbsolutePath VfsMountRootPath { get; }
```
Comment (narasamdya) [inline Public/Src/Cache/ContentStore/Vfs/VfsCasConfiguration.cs]: >VfsMountRootPath [](start = 28, length = 16)

Perhaps more comment on the difference between `VfsMountRootPath`, `VfsMountRelativeRoot` (why is it needed?), and `VfsRootPath`.

File: Public/Src/Cache/ContentStore/Vfs/VfsCasConfiguration.cs [hunk 5]
```diff
@@ -42,6 +42,26 @@ public class VfsCasConfiguration
         /// </summary>
         public AbsolutePath VfsRootPath { get; }
```
Comment (narasamdya) [inline Public/Src/Cache/ContentStore/Vfs/VfsCasConfiguration.cs]: >VfsRootPath [](start = 28, length = 11)

Is this still used?

File: Public/Src/Utilities/Native/IO/Windows/FileSystem.Win.cs [hunk 6]
```diff
@@ -387,7 +387,7 @@ private enum FileInfoByHandleClass : uint
         /// NotReadyDevice is treated as non-existent probe.
         /// BitLocker locked volume is treated as non-existent probe.
         /// </remarks>
-        public static bool IsHresultNonesixtent(int hr)
+        public static bool IsHresultNonexistent(int hr)
```
Comment (narasamdya) [inline Public/Src/Utilities/Native/IO/Windows/FileSystem.Win.cs]: >IsHresultNonexistent [](start = 27, length = 20)

I thought you don't like drama :grinning:

File: Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+
+namespace BuildXL.Cache.ContentStore.Vfs.Provider
+{
+    using Utils = Microsoft.Windows.ProjFS.Utils;
+
+    internal class ActiveEnumeration
```
Comment (narasamdya) [inline Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs]: >ActiveEnumeration [](start = 19, length = 17)

Can you add description what this class is for?

File: Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+using System.Collections.Generic;
+
+namespace BuildXL.Cache.ContentStore.Vfs.Provider
+{
+    using Utils = Microsoft.Windows.ProjFS.Utils;
+
+    internal class ActiveEnumeration
+    {
+        private readonly IEnumerable<VfsNode> fileInfos;
```
Comment (narasamdya) [inline Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs]: `_fileInfos` ?

File: Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        /// Advances the enumerator to the next element of the collection (that is being projected).   
+        /// If a filter string is set, MoveNext will advance to the next entry that matches the filter.
+        /// </summary>
+        /// <returns>
+        /// true if the enumerator was successfully advanced to the next element; false if the enumerator has passed the end of the collection
+        /// </returns>
+        public bool MoveNext()
+        {
+            IsCurrentValid = fileInfoEnumerator.MoveNext();
+            while (IsCurrentValid && IsCurrentHidden())
```
Comment (narasamdya) [inline Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs]: >IsCurrentHidden [](start = 37, length = 15)

`!FilterMatch()` ?

File: Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+            string filter)
+        {
+            if (string.IsNullOrEmpty(filter))
+            {
+                filterString = string.Empty;
+            }
+            else
+            {
+                filterString = filter;
+                if (IsCurrentValid && IsCurrentHidden())
```
Comment (narasamdya) [inline Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs]: >if [](start = 16, length = 2)

Why not loop until filter match?

File: Public/Src/Cache/ContentStore/Vfs/VfsTree.cs [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+            return null;
+        }
+    }
+
+    public abstract class VfsNode
+    {
+        public readonly string Name;
+        public readonly DateTime Timestamp;
+        public VfsNode NextSibling;
+        public VfsNode PriorSibling;
```
Comment (narasamdya) [inline Public/Src/Cache/ContentStore/Vfs/VfsTree.cs]: >Prior [](start = 23, length = 5)

Is this often called `PreviousSibling`?

File: Public/Src/Cache/ContentStore/Vfs/VfsTree.cs [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+using BuildXL.Cache.ContentStore.Interfaces.FileSystem;
+using BuildXL.Cache.ContentStore.Interfaces.Sessions;
+using BuildXL.Cache.ContentStore.Utils;
+using BuildXL.Utilities.Collections;
+
+namespace BuildXL.Cache.ContentStore.Vfs
+{
+    /// <summary>
+    /// A store which virtualizes calls to an underlying content store (i.e. content will
+    /// be lazily materialized using the projected file system filter driver)
```
Comment (narasamdya) [inline Public/Src/Cache/ContentStore/Vfs/VfsTree.cs]: Copy-paste?

File: Public/Src/Cache/ContentStore/Vfs/VirtualizedContentSession.cs [hunk 13]
```diff
@@ -24,89 +26,127 @@
...
+                }
+            }
+
+            var virtualPath = _contentManager.ToVirtualPath(path);
+            if (virtualPath == null)
+            {
+                return await _innerSession.PlaceFileAsync(operationContext, contentHash, path, accessMode, replacementMode, realizationMode, operationContext.Token, urgencyHint);
+            }
+
+            _contentManager.Tree.AddFileNode(virtualPath, new VfsFilePlacementData(contentHash, realizationMode, accessMode));
```
Comment (narasamdya) [inline Public/Src/Cache/ContentStore/Vfs/VirtualizedContentSession.cs]: >AddFileNode [](start = 33, length = 11)

So is this just placing the content in memory?

File: Public/Src/App/Bxl/Args.cs [hunk 14]
```diff
@@ -1071,6 +1071,9 @@ public bool TryParse(string[] args, PathTable pathTable, out ICommandLineConfigu
                         OptionHandlerFactory.CreateBoolOption(
                             "verifyCacheLookupPin",
                             sign => schedulingConfiguration.VerifyCacheLookupPin = sign),
+                        OptionHandlerFactory.CreateOption(
+                            "vfsCasRoot",
```
Comment (mpysson) [inline Public/Src/App/Bxl/Args.cs]: >vfsCasRoot [](start = 29, length = 10)

Add help text

File: Public/Src/Cache/ContentStore/Interfaces/FileSystem/VfsUtilities.cs [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+using BuildXL.Cache.ContentStore.Hashing;
+using BuildXL.Cache.ContentStore.Interfaces.Sessions;
+using BuildXL.Cache.ContentStore.Interfaces.Utils;
+
+namespace BuildXL.Cache.ContentStore.Interfaces.FileSystem
+{
+    /// <summary>
+    /// Defines utility methods for working with VFS
+    /// </summary>
+    public static class VfsUtilities
```
Comment (mpysson) [inline Public/Src/Cache/ContentStore/Interfaces/FileSystem/VfsUtilities.cs]: >VfsUtilities [](start = 24, length = 12)

Unit tests?

File: Public/Src/Cache/ContentStore/Vfs/VfsContentManager.cs [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+    using VirtualPath = System.String;
+
+    /// <summary>
+    /// A store which virtualizes calls to an underlying content store (i.e. content will
+    /// be lazily materialized using the projected file system filter driver)
+    /// </summary>
+    internal class VfsContentManager : IDisposable
+    {
+        // TODO: Track stats about file materialization (i.e. how much content was hydrated)
+        // On Domino side, track how much requested total requested file content size would be.
```
Comment (mpysson) [inline Public/Src/Cache/ContentStore/Vfs/VfsContentManager.cs]: >Domino [](start = 14, length = 6)

BuildXL :grinning:

File: Public/Src/Cache/ContentStore/Vfs/VfsTree.cs [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+        {
+            return _nodeMap.TryGetValue(relativePath, out node);
+        }
+
+        /// <summary>
+        /// Adds a file node (and any needed parent directory nodes) at the VFS root relative path
+        /// </summary>
+        public VfsFileNode AddFileNode(string relativePath, VfsFilePlacementData data)
+        {
+            var timestamp = DateTime.UtcNow;
```
Comment (mpysson) [inline Public/Src/Cache/ContentStore/Vfs/VfsTree.cs]: I presume this is arbitrary and Vfs doesn't care if it's representative. Is that correct?

File: Public/Src/Cache/ContentStore/Vfs/Provider/VfsProvider.cs [hunk 18]
```diff
@@ -0,0 +1,507 @@
...
+
+        private HResult QueryFileNameCallback(string relativePath)
+        {
+            // First try normal lookup
+            if (Tree.TryGetNode(relativePath, out var node))
+            {
+                return HResult.Ok;
+            }
+
+            if (!Utils.DoesNameContainWildCards(relativePath))
```
Comment (mpysson) [inline Public/Src/Cache/ContentStore/Vfs/Provider/VfsProvider.cs]: >Utils.DoesNameContainWildCards [](start = 17, length = 30)

Do these play nicely with long paths?

File: Public/Src/Cache/ContentStore/Interfaces/FileSystem/VfsUtilities.cs (hunk 2 above)
Comment (ljcollins25) [reply Public/Src/Cache/ContentStore/Interfaces/FileSystem/VfsUtilities.cs]: This is check. No need to mutate the string.

---
In reply to: [298305804](https://github.com/microsoft/BuildXL/pull/519#discussion_r298305804) [](ancestors = 298305804)

File: Public/Src/Cache/ContentStore/Vfs/VfsCasConfiguration.cs (hunk 5 above)
Comment (ljcollins25) [reply Public/Src/Cache/ContentStore/Vfs/VfsCasConfiguration.cs]: Yes, lots of times.

---
In reply to: [298309708](https://github.com/microsoft/BuildXL/pull/519#discussion_r298309708) [](ancestors = 298309708)

File: Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs (hunk 8 above)
Comment (ljcollins25) [reply Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs]: I'm leaving the styling mostly unchanged to ensure better comparison with original sample where this code came from.

---
In reply to: [298336392](https://github.com/microsoft/BuildXL/pull/519#discussion_r298336392) [](ancestors = 298336392)

File: Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs (hunk 10 above)
Comment (ljcollins25) [reply Public/Src/Cache/ContentStore/Vfs/Provider/ActiveEnumeration.cs]: Not sure. This is how the sample chose to do this. I don't want to change behavior to risk breaking things.

---
In reply to: [298338318](https://github.com/microsoft/BuildXL/pull/519#discussion_r298338318) [](ancestors = 298338318)

File: Public/Src/Cache/ContentStore/Vfs/VfsTree.cs (hunk 11 above)
Comment (ljcollins25) [reply Public/Src/Cache/ContentStore/Vfs/VfsTree.cs]: Not sure. They are equivalent as far as I'm concerned. I'll rename.

---
In reply to: [298340773](https://github.com/microsoft/BuildXL/pull/519#discussion_r298340773) [](ancestors = 298340773)

File: Public/Src/Cache/ContentStore/Interfaces/FileSystem/VfsUtilities.cs (hunk 15 above)
Comment (ljcollins25) [reply Public/Src/Cache/ContentStore/Interfaces/FileSystem/VfsUtilities.cs]: TBD. This is a prototype.

---
In reply to: [298425687](https://github.com/microsoft/BuildXL/pull/519#discussion_r298425687) [](ancestors = 298425687)

File: Public/Src/Cache/ContentStore/Vfs/VfsTree.cs (hunk 17 above)
Comment (ljcollins25) [reply Public/Src/Cache/ContentStore/Vfs/VfsTree.cs]: Correct. It's mainly whether a file system consumer cares.

---
In reply to: [298427084](https://github.com/microsoft/BuildXL/pull/519#discussion_r298427084) [](ancestors = 298427084)

File: Public/Src/Cache/ContentStore/Vfs/Provider/VfsProvider.cs (hunk 18 above)
Comment (ljcollins25) [reply Public/Src/Cache/ContentStore/Vfs/Provider/VfsProvider.cs]: Don't know. This is implemented in the native layer. 

---
In reply to: [298428409](https://github.com/microsoft/BuildXL/pull/519#discussion_r298428409) [](ancestors = 298428409)

File: Public/Src/Cache/ContentStore/Vfs/VirtualizedContentSession.cs (hunk 13 above)
Comment (ljcollins25) [reply Public/Src/Cache/ContentStore/Vfs/VirtualizedContentSession.cs]: Correct.

---
In reply to: [298369962](https://github.com/microsoft/BuildXL/pull/519#discussion_r298369962) [](ancestors = 298369962)

**Pull request merged**